### ISSUE TITLE
Thinking about creating my own launcher.

### ISSUE THREAD

**Issue by HunterArmstrong** [TBog/TBog#1]
Hey, I'm thinking about creating my own launcher for the first time and I'm thinking about using the KISS launcher as a starting off point. I'm just wondering what open source files did you use from the KISS GitHub to make the TB Launcher? Also, are there are any tips you want to give me for when I start my project?

**Comment by TBog** [comment]
What I did was to create a new project using an Android Studio template. After I had a working activity I started to import features from KISS.

Some features like application list loading and searching logic I imported as is.
For widgets, shortcuts and backup I implemented my own logic.

I also restructured the code to my liking and because of this I can't easily merge new features from KISS.

You can also just get all the code from KISS and just change the project name and path to start with a working launcher. My way of doing thinkgs is slow but I wanted my own launcher without depending on KISS for the interface and the result list.

**Comment by TBog** [comment]
Also, I wanted to use the androidx library and has a lot of work to do in order to configure the themes and Settings Activity.
Also because of this my apk size probably increased a lot.

**Comment by HunterArmstrong** [comment]
> You can also just get all the code from KISS and just change the project name and path to start with a working launcher. My way of doing thinkgs is slow but I wanted my own launcher without depending on KISS for the interface and the result list.

Where do you think I could get the KISS code from? Im thinking of just working with the code that allows you to search apps, contacts, shortcuts, and web results from KISS. 

I found a different type of launcher that is also open source with an interesting interface. I just want to replace the search bar function with the KISS search function because I find that there really isn't a way to customize the search bar or it's functions.

**Comment by TBog** [comment]
The search function of KISS is not easy to separate from the rest of the launcher, but I'll try to point you in the right direction when I get to my PC.

Imho it would be easier to separate it from my launcher.

**Comment by HunterArmstrong** [comment]
Well, in the meantime, I'll do my best with what I have.

**Comment by HunterArmstrong** [comment]
@TBog Hey, I am running into a small issue. The device I am running Android studio on cant emulate the app I am making in its emulator. Is there a way I can use a device I already have instead of the emulator?

**Comment by TBog** [comment]
> Is there a way I can use a device I already have instead of the emulator?

Not sure about windows, but I think you need a driver. 
You should just need to plug the device in the USB and Android Studio will detect it. Then you'll run the app on the device.

**Comment by HunterArmstrong** [comment]
Oh Okay. Does the BlueStacks emulator work as well?

**Comment by TBog** [comment]
> Oh Okay. Does the BlueStacks emulator work as well?

Never used that.
Afaik you can debug only using the device or Android Studio's emulator

**Comment by HunterArmstrong** [comment]
Okay. Well, I guess I am just having issues with the emulator. 

Whenever I am trying to use a virtual device, I get this red error text that reads "Android Emulator Hypervisor Driver for AMD Processors is not installed" 

Do I have to install another app on my computer or is there just a random bug?

**Comment by TBog** [comment]
use Google
I found this:
https://androidstudio.googleblog.com/2019/10/android-emulator-hypervisor-driver-for.html?m=1

**Comment by HunterArmstrong** [comment]
Okay. Ill see if i can get the emulator up and running. 

I also want to start my project with the launcher I found as a starting off point. Is there a way to easily transfer coding for the launcher from github/google into Android Studio as a new project?

**Comment by TBog** [comment]
Both KISS and TBLauncher can be opened with Android Studio as is.

**Comment by HunterArmstrong** [comment]
Okay. Cool. I'm just wondering, was there any tutorial videos or a course that helped you learn how to use Android Studio easily.

**Comment by TBog** [comment]
Sorry, I just opened it and it mostly worked on its own.
Any issue I have I google it.

**Comment by HunterArmstrong** [comment]
Cool! 

Is there a way to make the launcher an overlay?

**Comment by TBog** [comment]
> Is there a way to make the launcher an overlay?

Of course. You probably need the appropriate permissions. Also you should read the documentation and try it out on a smaller scale and check an example. 
I have no experience with this feature and can't help more. 

Here is how to ask the user for the permission: https://stackoverflow.com/questions/40355344/how-to-programmatically-grant-the-draw-over-other-apps-permission-in-android

here is an example: https://www.google.com/amp/s/mobikul.com/how-to-show-screen-overlay-in-android/amp/

**Comment by 8singularity** [comment]
@HunterArmstrong, don't mean to hijack your chat, buy can you name the "the launcher I found as a starting off point"? I'm also looking into making my own launcher, so that other launcher may be helpful.

@TBog, big thank you for TBLauncher!  Up to now, my essential launcher has been Fastkey Launcher; nothing else I found could match it, but that's been stagnant since 2018. Now I see that your TBLauncher may be what I've been looking for! Thanks again!

**Comment by TBog** [comment]
@8singularity Why are you looking to make your own? Is there a feature you'd like to see?

**Comment by 8singularity** [comment]
@TBog thanks for responding.  
I am a noob at programming, so my intent was to learn as I go.  The primary goal is to learn by doing and failing, so I don't expect to actually have a working launcher.  Since you are asking, I would like a few tweaks and changes, which would at least be helpful for me.  An adjustable font size in the search bar, possibly autoadjusted based on the height of the search bar.  (while poking around, I just realized I can do this.  Ergo, a better organized menu, maybe?)  A swipeable keyboard that autolaunches when the app when the results are narrowed to one (like Fast App Search Tool).  Some documentation with some examples of use case would be helpful.  2 or 3 presets would be very nice.  :)

**Comment by TBog** [comment]
Why not take an open source launcher (like [TBLauncher](https://tbog.github.io/TBLauncher/), or [KISS](https://kisslauncher.com/), or whatever else you find) and contribute your changes. If the maintainer disagrees you can fork from that or keep your branch and always merge with the main branch when a new release is ready.
I think this is the best way to learn.
If you start your launcher you will not have a working one for a while and nobody will be checking your code for bugs.

**Comment by 8singularity** [comment]
Oh yes, I agree. I intend to do just that, but I wouldn't be of any help on the coding side, which is why I asked for existing launcher codes that I can examine. 

I've been digging into your launcher, as a user and not your code, trying to figure out and understand how to use it. You've built a LOT of functionality into it. I think it pretty much has everything I've been asking for, and a whole lot more! Do you have our plan to write an instructional manual or video for TBLauncher? There are so many features hidden or nested away, that a manual on the basics and would be super helpful.

Lynx launcher and Sentient  launcher are two similar launders. But for now, TDLauncher is my fauldaily driver, and I loving it!
